File: Common/Servers/LocalOscillator/src/CommandLine.cpp
/*******************************************************************************
* ALMA - Atacama Large Millimiter Array
* (c) UNSPECIFIED - FILL IN, 2011 
* 
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the License, or (at your option) any later version.
* 
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* Lesser General Public License for more details.
* 
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
*
* "@(#) $Id$"
*
* who       when      what
* --------  --------  ----------------------------------------------
* spoppi  2011-09-26  created 
*/

/************************************************************************
*   NAME
*   
* 
*   SYNOPSIS
*
*   
*   PARENT CLASS
*
* 
*   DESCRIPTION
*
*
*   PUBLIC METHODS
*
*
*   PUBLIC DATA MEMBERS
*
*
*   PROTECTED METHODS
*
*
*   PROTECTED DATA MEMBERS
*
*
*   PRIVATE METHODS
*
*
*   PRIVATE DATA MEMBERS
*
*
*   FILES
*
*   ENVIRONMENT
*
*   COMMANDS
*
*   RETURN VALUES
*
*   CAUTIONS 
*
*   EXAMPLES
*
*   SEE ALSO
*
*   BUGS   
* 
*------------------------------------------------------------------------
*/

#include "vltPort.h"
#include "CommandLine.h"
static char *rcsId="@(#) $Id$"; 
static void *use_rcsId = ((void)&use_rcsId,(void *) &rcsId);


static const int MAXLEN=255; // max len of the reply from a gpib quert


static const string FREQCMD="FREQ ";
static const string QUERYFREQ="FREQ? ";
static const string QUERYPOWER="POWER? ";
static const string POWERCMD="FREQ AMPL";
static const string RFONCMD="OUTP:STAT ON";
static const string RFOFFCMD="OUTP:STAT OFF";
static const string QUERYRF="OUTP:STAT?";
static const string FREQUNIT=" MHZ";
static const string POWERUNIT=" dBM";



CommandLine::CommandLine()
{
//	ACS_TRACE("CommandLine::CommandLine()");
	GPIB_errors[EDVR]="System error";
	GPIB_errors[ECIC]="Function requires GPIB board to be CIC";
    GPIB_errors[ENOL]="No Listeners on the GPIB";
    GPIB_errors[EADR]="GPIB board not addressed correctly";
    GPIB_errors[EARG]="Invalid argument to function call";
    GPIB_errors[ESAC]="GPIB board not System Controller as required";
    GPIB_errors[EABO]="I/O operation aborted (timeout)";
    GPIB_errors[ENEB]="Nonexistent GPIB board";
    GPIB_errors[EDMA]="DMA error";
    GPIB_errors[EOIP]="Asynchronous I/O in progress";
    GPIB_errors[ECAP]="No capability for operation";
    GPIB_errors[EFSO]="File system error";
    GPIB_errors[EBUS]="GPIB bus error";
    GPIB_errors[ESRQ]="SRQ stuck in ON position";
    GPIB_errors[ETAB]="Table problem";
    GPIB_errors[ELCK]="Interface is locked";
    GPIB_errors[EARM]="Ibnotify callback failed to rearm";
    GPIB_errors[EWIP]="Wait in progress on specified input handle";
    GPIB_errors[ERST]="Event notification was cancelled due to a reset of the interface";
    GPIB_errors[EPWR]="The interface lost power";

}

CommandLine::~CommandLine()
{


}


int CommandLine::configure(int bid,int pad,int sat,int tmo, int eot, int eos) throw(GPIBException)

{

	m_device = ibdev(                /* Create a unit descriptor handle         */
	                    bid,              /* Board Index (GPIB0 = 0, GPIB1 = 1, ...) */
	                    pad,          /* Device primary address                  */
	                    sat,        /* Device secondary address                */
	                    tmo,                    /* Timeout setting (T10s = 10 seconds)     */
	                    eot,                       /* Assert EOI line at end of write         */
	                    eos);

 	if (m_device==ERR) {
 						cout <<"err" <<endl;
					   throw GPIBException("ibdev:"+GPIB_errors[ThreadIberr()]);
					}
	if (m_device==TIMO) {
					   throw GPIBException("ibdev:timeout");
					}
	if (ThreadIbsta()& ERR )
		{
						   throw GPIBException("ibdev:"+GPIB_errors[ThreadIberr()]);
		}
	return 0;

}

int CommandLine::connect(string& reply)
{
	return 0;


}

int CommandLine::query(string cmd,string& reply) throw (GPIBException)
{

	char buff[MAXLEN];

	ibwrt(m_device,(char* )cmd.c_str(), cmd.size());
	if (ThreadIbsta()& ERR )
	{
 					   throw GPIBException("ibrwt:"+GPIB_errors[ThreadIberr()]);
	}
/* Send the identification query command   */
	ibrd(m_device, buff, MAXLEN);
	if (ThreadIbsta()& ERR )
		{
						   throw GPIBException("ibrd:"+GPIB_errors[ThreadIberr()]);
		}/* Read up to 100 bytes from the device
	*/
	buff[ThreadIbcnt()] = '\0';	     /* Null terminate the ASCII string 	*/
	reply = string(buff) ;

	return 0;


}
int CommandLine::disconnect() throw (GPIBException)
{
	ibonl(m_device,0);
	if (ThreadIbsta()& ERR )
		{
			throw GPIBException("ibonl:"+GPIB_errors[ThreadIberr()]);
		}

	return 0;

}

int CommandLine::init(string& reply) throw (GPIBException)
		{
			try {
				query("*IDN?",reply);
				sendCMD("*RST"); // reset
				sendCMD("POW:ALC:SOUR INT"); // enable internal leveling;
			}  	catch (GPIBException& ex)

			{
					throw ex;
			}
			return 0;

		}



int CommandLine::setPower(double pow) throw (GPIBException)
{

	try{
		string cmd;
		cmd="POW " + stringify(pow)+POWERUNIT;
		sendCMD(cmd);

	} catch (GPIBException& ex)

	{

		throw;

	}
	return 0;

}

int CommandLine::getFreq(double & freq) throw (GPIBException)

{
	try {
		string reply;
		query(QUERYFREQ,reply);
		cout << "commandline freq:" <<reply << endl;
		freq=atof(reply.c_str());
		return 0;

	}  	catch (GPIBException& ex)

	{


			throw ex;
			return -1;

	}
}


int CommandLine::setFreq(double freq) throw (GPIBException){

	try{

		string cmd;
		cmd=FREQCMD+ stringify(freq)+FREQUNIT;
		sendCMD(cmd);

	} catch (GPIBException& ex)

	{


		throw ex;

	}


	return 0;




}

int CommandLine::getPower(double& pow)  throw (GPIBException)
{
	try{
		        string reply;
				query(QUERYPOWER,reply);
				pow=atof(reply.c_str());
	}catch(GPIBException& ex){
		throw ex;
	}
	return 0;

}

int CommandLine::sendCMD(string cmd) throw(GPIBException)
{

 	ibwrt(m_device,const_cast<char*>(cmd.c_str()), cmd.size());
	int i;
	i=ThreadIbsta();

 	if (i& ERR )
	{
		cout << "Error:" <<endl;
		throw GPIBException("ibrwt:"+GPIB_errors[ThreadIberr()]);
	}

	return 0;

}


void CommandLine::queryErrors(string& reply) throw(GPIBException)
{

	try{
	  //      string reply;
			query("SYST:ERR?",reply);

 		}catch(GPIBException& ex){
			throw ex;
		}
 }

int CommandLine::parseSCPIErrors(string msg,string& reply)

{

	// the scpi errors, i.e. the internal errors queue, is
	// made of   code,"message" pair
	// e.g.   -113,"Undefined header"

	int code;

	size_t comma;
	comma=msg.find(",") ; //
	code =atoi( msg.substr(0,comma).c_str());
	reply=msg.substr(comma+1);
	return code;




}

int CommandLine::rfOn() throw (GPIBException)
{
	try{
		string reply;
		int iReply;

		query(QUERYRF,reply);
		iReply=atof(reply.c_str());
		if (iReply == 0){
			sendCMD(RFONCMD);

		} //  turn on only if is off.


	} catch (GPIBException& ex)
	{
		throw ex;

	}



}

int CommandLine::rfOff() throw (GPIBException)
{

	try{
			string reply;
			int iReply;
			query(QUERYRF,reply);
			iReply=atof(reply.c_str());

			if (iReply == 1){
				sendCMD(RFOFFCMD);

			} //  turn on only if is off.


		} catch (GPIBException& ex)
		{
			throw ex;

		}



}

/*___oOo___*/

File: Noto/Servers/LocalOscillator/src/LocalOscillatorImpl/CommandLine.py
import socket
import time
import decimal

QUERYERROR="SYST:ERR? \n"

FREQCMD="FREQ "
QUERYFREQ="FREQ?;"+QUERYERROR
QUERYPOWER="POW?\n"
RFONCMD="OUTP:STAT ON"
RFOFFCMD="OUTP:STAT OFF"
QUERYRF="OUTP:STAT?"
FREQUNIT=" MHZ\n"
POWERUNIT=" dBM\n"

class CommandLineError(Exception):
   def __init__(self, value):

      self.value = value
   def __str__(self):
      return repr(self.value)


class CommandLine:
   
    
   def __init__(self):
     
       try: 
          self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
       except socket.error , msg:
          print msg
          

   def __del__(self):
       pass
   
   def configure(self,ip,port):
      '''
      Connect to the HW
      Clear query error
      
      '''
        
   
      try:
         self.ip=ip
         self.port=port
         
         self.sock.connect((ip,port))
         msg_ok ='OK' 
         self.sock.sendall('*CLS\n;SYST:ERR?\n++read\n')
         msg = self.sock.recv(1024)

         return msg_ok
      except socket.error , msg:
         print msg
         print "connect error: " ,msg
         return msg
   
   def init(self,reply):
     
     pass
   
   def setPower(self,power):
       cmd= "POW " + str(power) + "DBM;SYST:ERR?\n"
 
       try:
           msg=self.query(cmd)   

           return  msg
           
       except socket.error , msg:
          print "connect error: " ,msg
          return msg
   
   def getPower(self):
       QUERYPOWER="POW?;SYST:ERR?\n"
       cmd=QUERYPOWER
       try:
          msg=self.query(cmd)
          commands=msg.split(';')
          
          if len(commands)>1:
             val=int(decimal.Decimal(commands[0]))# unit is MHZ,
             err_msg=commands[1]
          else:
             val=-1
             err_msg='Communication Error with synth'
           
          return err_msg,val
   
       except socket.error , msg:
          print "connect error: " ,msg
          return msg,-1
       except CommandLineError,msg:
          raise
       except ValueError,msg:
          raise CommandLineError(msg)   
   
     
   def setFrequency(self,freq):

       cmd=  'FREQ '+str(freq)+ 'MHZ;SYST:ERR?'

 
       try:
           msg=self.query(cmd)
           return msg
           
       except socket.error , msg:
          print "connect error: " ,msg
          return msg
   
   def getFrequency(self):
         
        cmd= 'FREQ?;SYST:ERR?'
        try:
           
          msg=self.query(cmd)
          commands=msg.split(';')
          if len(commands)>1:
              val=int(decimal.Decimal(commands[0]))/1e6 # unit is MHZ,
              err_msg=commands[1]
          else:
              val=-1
              err_msg='Communication Error with synth'
          print "Error Messager from Synt:",msg
          #if err_msg != '0,\"No error\"\n': 
                #print "exception",err_msg
                #raise CommandLineError(err_msg)   
          return err_msg,val
   
        except socket.error , msg:
          print "connect error: " ,msg
          return msg,-1
          
        except CommandLineError,msg:
          raise
        except ValueError,msg:
          raise CommandLineError(msg)   
   
   def readStatus(self):
       ''' 
       Query the error code of the synt.      
       '''       
       try:
           
          msg=self.query(QUERYERROR)
          print "query err",msg
          if msg != '0,\"No error\"\n': 
                print "exception",msg
                raise CommandLineError(msg)   
          return msg
       except socket.error , msg:
          print "connect error: " ,msg
          return msg
          
   def rfOn(self):
     
       pass
   
   def rfOff(self):
      
      
      pass
      
   def close(self):
     
       self.sock.close()

   def query(self,cmd):
       try:
           self.sock.sendall(cmd+'\n++read\n') 
           msg = self.sock.recv(1024) 
           if len(msg)==0:
             
                 self.sock.close()
                 
                 self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

                 self.configure(self.ip,self.port)
                 
           return msg
       
       except socket.error , msg:
          print "connect error: " ,msg
          print 'len msg received',len(msg)
          
          self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
          self.configure(self.ip,self.port)
  
          
          return msg
   
       
       
      
     